### PR TITLE
Embedded msg

### PR DESCRIPTION
**Pull request by devcsomnicg**
## JIRA Ticket(s) if any

* [DEV-233](https://iterable.atlassian.net/browse/DEV-233)

## Description
Module is all about to offers developers a set of methods and functionalities to efficiently handle messages sent from external sources 
and provides mechanisms to respond to updates and actions associated with these messages

## Release Notes
- improved coding standards
- improved type checks
- small bug fixes

### CODE REVIEW

File: src/embedded/embeddedManager.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+  public addActionHandler(actionHandler: EmbeddedMessageActionHandler) {
+    this.actionListeners.push(actionHandler);
+  }
+
+  // private notifyUpdateDelegates() {
+  //     this.updateListeners.forEach((updateListener: EmbeddedMessageUpdateHandler) => {
+  //         updateListener.onMessagesUpdated();
+  //     });
+  // }
```
Comment (mprew97) [inline src/embedded/embeddedManager.ts:76]: Is this needed or no?

File: src/embedded/embeddedManager.ts [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  public async syncMessages(userId: string, callback: () => void) {
+    await this.retrieveEmbeddedMessages(userId);
+    callback();
+  }
+
+  private async retrieveEmbeddedMessages(userId: string) {
+    try {
+      const iterableResult: any = await baseIterableRequest<IterableResponse>({
+        method: 'GET',
+        url: '/embedded-messaging/messages?userId=' + userId
```
[the remark this answers is not in the document]
Comment (mprew97) [reply src/embedded/embeddedManager.ts:25]: `endpoints` and their respective URLs should live in a `const` file somewhere and be imported. could probably just create a `consts.ts` file in this dir.

File: src/events/events.ts [hunk 3]
```diff
@@ -19,6 +19,46 @@ export const track = (payload: InAppTrackRequestParams) => {
...
+    validation: {
+      data: trackEmbeddedMessageSchema
+    }
+  });
+};
+
+export const trackEmbeddedMessageClick = (payload: IEmbeddedMessageMetadata, buttonIdentifier: string, clickedUrl: string, appPackageName: string) => {
+  return baseIterableRequest<IterableResponse>({
+    method: 'POST',
+    url: '/embedded-messaging/events/click',
```
Comment (mprew97) [inline src/events/events.ts:36]: url value should be in a const

File: src/embedded/embeddedSessionManager.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        this.session.end = new Date();
+
+        if (!this.session.impressions?.length) {
+            console.log("No impressions in the session. Skipping tracking.");
+            return;
+        }
+
+        if(this.impressions.size) {
+            //reset session for next session start
+            this.session = new EmbeddedSession(undefined, undefined, "0", undefined);
```
Comment (mprew97) [inline src/embedded/embeddedSessionManager.ts:69]: This `reset` seems weird to me... feels like EmbeddedSession should have it's own `reset` method on the class or something.

**Review by mprew97 — CHANGES_REQUESTED** [review]

**Comment by mprew97** [comment]
Could we also pull all the events and type data under `embedded/events` or `embedded/types`? Those types and events files are unwieldy as is so would prefer to start separating them by message channel now that the web sdk will be used for more than just `in-app`

File: react-example/src/views/Embedded.tsx [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+          displayCount: 3,
+          duration: 10,
+        },
+        {
+          messageId: 'def456',
+          displayCount: 2,
+          duration: 8,
+        },
+      ],
+    };
```
[the remark this answers is not in the document]
Comment (devcsomnicg) [reply react-example/src/views/Embedded.tsx:94]: Yes, it will come later

File: src/embedded/embeddedSessionManager.ts (hunk 4 above)
Comment (devcsomnicg) [reply src/embedded/embeddedSessionManager.ts:69]: It's re-initialising EmbeddedSession instance with some undefined values, i have an updated comment, in my sense there is no need of reset method separately.

File: src/events/types.ts [hunk 6]
```diff
@@ -24,3 +24,61 @@ export interface InAppEventRequestParams {
...
+  
+  buttons?: [IEmbeddedMessageElementsButton]
+  text?: [IEmbeddedMessageElementsText]
+  defaultAction?: IEmbeddedMessageElementsDefaultAction
+}
+
+export interface IEmbeddedMessage {
+  metadata: IEmbeddedMessageMetadata
+  elements?: IEmbeddedMessageElements
+  payload?: Array<any>
```
[the remark this answers is not in the document]
Comment (devcsomnicg) [reply src/events/types.ts:69]: I have taken reference of swift-sdk while preparing embedded message module, where payload is array of hashtable, so in js we can make Array<any> for now

Ref: https://github.com/Iterable/swift-sdk/blob/embedded-messaging/swift-sdk/IterableEmbeddedMessage.swift#L11

File: react-example/src/views/Embedded.tsx [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+  useEffect(() => {
+    initialize(process.env.API_KEY);
+  }, []);
+
+  const handleFetchEmbeddedMessages = async (e: FormEvent<HTMLFormElement>) => {
+    e.preventDefault();
+    setFetchingEmbeddedMessages(true);
+    try {
+      await new EmbeddedManager().syncMessages('harrymash2006', () =>
```
[the remark this answers is not in the document]
Comment (devcsomnicg) [reply react-example/src/views/Embedded.tsx]: This is removed, and added input box to enter client user id

File: src/embedded/embeddedManager.ts (hunk 1 above)
Comment (devcsomnicg) [reply src/embedded/embeddedManager.ts]: Removed commented code

File: react-example/src/views/Embedded.tsx [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+      </EndpointWrapper>
+
+      <Heading>POST /embedded-messaging/events/received</Heading>
+      <EndpointWrapper>
+        <Form onSubmit={submitEmbeddedMessagesReceivedEvent} data-qa-cart-submit>
+          <Button type="submit">
+            Submit
+          </Button>
+        </Form>
+      </EndpointWrapper>
```
[the remark this answers is not in the document]
Comment (devcsomnicg) [reply react-example/src/views/Embedded.tsx]: Applied suggested approach

File: src/embedded/embeddedManager.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (devcsomnicg) [reply src/embedded/embeddedManager.ts]: constants are moved in consts.ts file

File: src/events/events.ts (hunk 3 above)
Comment (devcsomnicg) [reply src/events/events.ts]: done

**Comment by devcsomnicg** [comment]
> Could we also pull all the events and type data under `embedded/events` or `embedded/types`? Those types and events files are unwieldy as is so would prefer to start separating them by message channel now that the web sdk will be used for more than just `in-app`

@mprew97 Can you explain more like, by pointing to file or specific methods and how you want to re-arrange or improve?

**Comment by mprew97** [comment]
> > Could we also pull all the events and type data under `embedded/events` or `embedded/types`? Those types and events files are unwieldy as is so would prefer to start separating them by message channel now that the web sdk will be used for more than just `in-app`
> 
> @mprew97 Can you explain more like, by pointing to file or specific methods and how you want to re-arrange or improve?

Of course! 

We have the current structure `src/events/`. Within that folder, we have 2 files that i think would better be suited to live under separated directories.  

So for example, I'm thinking that all of the embedded types in `src/events/types` should live in a new directory at `src/events/embedded/types` and the embedded events in `src/events/events` should live in `src/events/embedded/events`. And then we should go ahead and do the same for inapp `src/events/in-app/types` and `src/events/in-app/events`.

**Review by mprew97 — APPROVED** [review]

**Pull request merged**